sssd_test_framework/utils/dbus_types.py

```diff
@@ -0,0 +1,675 @@
+"""D-Bus Type mapping to and from Python Types"""
+
+from __future__ import annotations
+
+import re
+from abc import ABC, abstractmethod
+
+__all__ = [
+    "DBUSResult",
+    "DBUSType",
+    "DBUSTypeBoolean",
+    "DBUSTypeString",
+    "DBUSTypeObjectPath",
+    "DBUSTypeInteger",
+    "DBUSTypeByte",
+    "DBUSTypeInt16",
+    "DBUSTypeInt32",
+    "DBUSTypeInt64",
+    "DBUSTypeUInt16",
+    "DBUSTypeUInt32",
+    "DBUSTypeUInt64",
+    "DBUSTypeDouble",
+    "DBUSTypeContainer",
+    "DBUSTypeArray",
+    "DBUSTypeDict",
+    "DBUSTypeVariant",
+    "DBUSSignatureReader",
+]
+
+
+# A NOTE ABOUT DBUS-SEND
+# **********************
+# It is important to know that dbus-send does not have a symmetric
+# behavior with relation to how the types are described in text
+# form when used as parameters or results.
+
+
+# TYPES PENDING IMPLEMENTATION
+# ****************************
+# The following D-Bus types are not implemented because they are not
+# currently recognized by dbus-send:
+# - Signature
+# - UNIX FD
+# - Struct
+
+
+class MutableString:
+    """A helper class implementing an extremely simple mutable string"""
+
+    def __init__(self, text: str):
+        self.text = text
+
+    def __str__(self):
+        return self.text
+
+
+class DBUSResult(MutableString):
+    """
+    The result from the D-Bus operation executed by dbus-send.
+    It is used to simplify parsing of the result as the string
+    must be modified and have no carriage return.
+    """
+
+    def __init__(self, text: str):
+        self.text = text.replace("\n", "")
+
+
+class DBUSType(ABC):
+    """
+    Abstract class to create classes that handle type conversion
+    between D-Bus and Python.
+    """
+
+    # It would be good to define here the following abstract class members:
+    #    _type_prefix
```

Comment:
> Abstract class is there to define what HAS to be implemented by inherited classes and to ENFORCE it to avoid bugs. This does not enforce anything, it does not help.

I know very well what an abstract class is. That's why I added that comment explaining why I didn't do it as I would had liked to. The idea of this comment is to let other developers know it when they wonder "why did he do it this way and not this other way?."

> Set it in the constructor.

I considered this, but that would make the variables instance variables instead of class variables.

> Make it an abstract property

Yes, that's what I said in my comment. But I find it overkill to implement all that, considering the alternative solution I used also works and that none of the solutions is perfect. Mine doesn't enforce the class variables, but is simpler to implement and maintain; specially considering that the chances of D-Bus adding a new integer type and us implementing a new subclass of `DBUStypeInt` are very low.

> There is no reason to use class members here and something like "abstract class member" does not exist because it does not make any sense.

Oh. I strongly disagree with this. I see plenty of reasons for it to exist. This  use case is one of them.

> Class member is a class member, it it accessible only via the class name,

And this seem to be the mother of the problem. Python behaves in this strange way. Normally, a class member is accessible through the class and through an object, and it always remains a class member. Python behaves differently and it is not going to change any time soon, so we have no choice but to deal with it.

Reply:
Ok, let's agree to disagree.

> > Set it in the constructor.
> 
> I considered this, but that would make the variables instance variables instead of class variables.
> 
> > Make it an abstract property
> 
> Yes, that's what I said in my comment. But I find it overkill to implement all that, considering the alternative solution I used also works and that none of the solutions is perfect. Mine doesn't enforce the class variables, but is simpler to implement and maintain; specially considering that the chances of D-Bus adding a new integer type and us implementing a new subclass of `DBUStypeInt` are very low.

Ok. I don't agree with you and I don't like it, but feel free to make it the way you feel is right.

Reply:
I'm trying to implement the property as you suggested, but I'm facing some problems related to Python. For instance, I can define the property in the ``DBUSType`` class as a concrete inheritable method:
```
@property
def type_prefix(self) -> str:
    return self._type_prefix
```
In this case the variable ``_type_prefix``, which is declared in the concrete classes, will not be recognized by this abstract class and I will have to ignore the error. This is what we are trying to avoid.

Or I can make the method abstract and not use this variable:
```
@property
@abstractmethod
def type_prefix(self) -> str:
    pass
```
And every single concrete class will have to implement this method (which is right) but all the implementations will be identical, replicating the code above on every concrete class.

Is there a third option? Which is the pythonic way of doing this?

Reply:
Hmm, I don't think I understand the problem.

The first approach:
```
@property
def type_prefix(self) -> str:
    return self._type_prefix
```
... is in my opinion still wrong, since you expect that `_type_prefix` is set in the child classes and it is pretty much the same as what you have now. You are creating requirement that is not checked. Python allows this. That's why we use tools like mypy to detect these errors to avoid runtime bugs.

The second approach is correct, but I'm not sure I understand `but all the implementations will be identical, replicating the code above on every concrete class`. Do you mean that every concrete class will have to implement this property but with different content? So you have issue with the fact that every class will contain the following three lines just with different string?

```
@property
def type_prefix(self) -> str
    return "the-concrete-class-prefix"
```

Let me know if I misunderstood you, otherwise this is the correct approach, there is no shortcut. Other way would be to use the constructor. Either way, we get static analyzers return an error if this property is not defined in the concrete classes.

It would be also possible to write your own metaclass and check that this field is set, essentially implementing your own abstract class variable, but that would be overthinking it.

---

But really, if you feel strongly about the current implementation, I am not going to block it.

Reply:
During the week-end I found the third method I was looking for. It happened to be what you propose in this message.

I was returning in the property the value of the class variable `_type_prefix`, which produced generic code which could be inherited but had the problem of not be accepted by linters. Then I noticed I didn't need that variable any more and each class could return its value directly. Code is not shared but the variable is no longer needed.